File: ProductService/src/main/java/com/cgi/productservice/Services/ProductService.java
package com.cgi.productservice.Services;

import com.cgi.productservice.Exceptions.ProductWithTheIDAlreadyExistException;
import com.cgi.productservice.Exceptions.ProductWithTheIDDoesNotExistException;
import com.cgi.productservice.Model.Product;


import java.util.List;

public interface ProductService {
    List<Product> getAllProducts();
    Product getProductById(Long id) throws ProductWithTheIDDoesNotExistException;
    Product addProduct(Product product) throws ProductWithTheIDAlreadyExistException;
    void deleteProductById(Long id) throws ProductWithTheIDDoesNotExistException;
    Product updateProduct(Product product) throws ProductWithTheIDAlreadyExistException;

}

File: Capstone/src/app/services/product.service.ts
import { Injectable } from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {Product} from "../model/product";

@Injectable({
  providedIn: 'root'
})
export class ProductService {
  private baseUrl="http://localhost:8080/api/v1/products";

  constructor(private httpClient: HttpClient) { }

  getAllProducts(): Observable<any[]> {
    return this.httpClient.get<any[]>("http://localhost:8080/api/v1/product");
  }

  addNewProduct(product: Product ): Observable<Product>{
    return this.httpClient.post<Product>(`http://localhost:8080/api/v1/product`, product)
  }

  updateProduct(product: Product){
    return this.httpClient.put<Product>("http://localhost:8080/api/v1/product/"+product.id, product)
  }

  getProductById(id:number):Observable<Product> {
    return this.httpClient.get<Product>(`${this.baseUrl}/${id}`);
  }

  deleteProduct(product: Product){
    return this.httpClient.delete<Product>("http://localhost:8080/api/v1/product/"+product.id);
  }

}
